cumulusci/tasks/package_2gp.py

```diff
@@ -0,0 +1,537 @@
+        )
+        self.logger.info(f"  Version Number: {self.return_values['version_number']}")
+        self.logger.info(f"  Dependencies: {self.return_values['dependencies']}")
+
+    def _get_or_create_package(self, package_config: PackageConfig):
+        """Find or create the Package2
+
+        Checks the Dev Hub for an existing, non-deprecated 2GP package
+        with matching name, type, and namespace.
+        """
+        message = f"Checking for existing {package_config.package_type} Package named {package_config.name}"
+        query = f"SELECT Id FROM Package2 WHERE IsDeprecated = FALSE AND ContainerOptions='{package_config.package_type}' AND Name='{package_config.name}'"
+        if package_config.namespace:
+            query += f" AND NamespacePrefix='{package_config.namespace}'"
+            message += f" with namespace {package_config.namespace}"
+        else:
+            query += " AND NamespacePrefix=null"
+        self.logger.info(message)
+        try:
+            res = self.tooling.query(query)
+        except SalesforceMalformedRequest as err:
+            if "Object type 'Package2' is not supported" in err.content[0]["message"]:
+                raise TaskOptionsError(
+                    "This org does not have a Dev Hub with 2nd-generation packaging enabled. "
+                    "Make sure you are using the correct org and/or check the Dev Hub settings in Setup."
+                )
+        if res["size"] > 1:
+            raise TaskOptionsError(
+                f"Found {res['size']} packages with the same name, namespace, and package_type"
+            )
+        if res["size"] == 1:
+            package_id = res["records"][0]["Id"]
+            self.logger.info(f"Found {package_id}")
+            return package_id
+
+        self.logger.info("No existing package found, creating the package")
+        Package2 = self._get_tooling_object("Package2")
+        package = Package2.create(
+            {
+                "ContainerOptions": package_config.package_type,
+                "Name": package_config.name,
+                "Description": package_config.description,
+                "NamespacePrefix": package_config.namespace,
+            }
+        )
+        return package["id"]
+
+    def _create_version_request(self, package_id, package_config, package_zip_builder):
+        # Prepare the VersionInfo file
+        version_bytes = io.BytesIO()
+        version_info = zipfile.ZipFile(version_bytes, "w", zipfile.ZIP_DEFLATED)
+        try:
+
+            # Add the package.zip
+            package_hash = package_zip_builder.as_hash()
+            version_info.writestr("package.zip", package_zip_builder.as_bytes())
+
+            # Check for an existing package with the same contents
+            res = self.tooling.query(
+                "SELECT Id "
+                "FROM Package2VersionCreateRequest "
+                f"WHERE Package2Id = '{package_id}' "
+                "AND Status != 'Error' "
```

Comment:
We should add a where clause for SkipValidation here since if you want to promote a package that had been previously uploaded with SkipValidation you'd need to re-upload the package with SkipValidation set to false

Reply:
And writing this down because we talked about it but I haven't done yet: re-add the option to force re-uploading.